### PR TITLE
[Soc2014] Official meta API

### PR DESCRIPTION
**Pull request by PirosB3**

### CODE REVIEW

File: django/db/models/options.py [hunk 1]
```diff
@@ -309,253 +363,218 @@ def _swapped(self):
...
+            # Due to the way Django's internals work, get_field() should also be able to fetch a field by attname.
+            # Which in the case of a concrete field with relation, includes the *_id name too
+            try:
+                res[field.attname] = field
+            except AttributeError:
+                pass
+
+        return res
+
+    def get_field(self, field_name, **kwargs):
```
Comment (freakboy3742) [inline django/db/models/options.py]: Not sure where "above" is, but I agree this change should be made.

File: django/db/models/fields/related.py [hunk 2]
```diff
@@ -98,6 +98,13 @@ def do_pending_lookups(sender, **kwargs):
 
 
 class RelatedField(Field):
+    has_relation = True
```
Comment (freakboy3742) [inline django/db/models/fields/related.py]: This is redundant now; related fields will get this automatically as a result of having a "rel" attribute.

**Comment by freakboy3742** [comment]
I've just completed a survey of all the outstanding comments on this PR, creating a TODO list for merge:

- [ ] Revert import reorganisation in django/contrib/admin/checks.py
- [ ] Remove redundant has_relation from django/db/models/fields/related.py
- [ ] Cleaned up docstring text in django/db/models/options.py
- [ ] Refactor get_field() to use explicit many_to_many argument.
- [ ] Resolve use of local_* in django/db/models/base.py
- [ ] Resolve bad reference in releases/1.8.txt, 
- [ ] Resolve use of local_fields etc.
- [ ] Add migration guide to releases/1.8.txt
- [ ] Add tests for new field types (UUIDField, etc)
- [ ] Resolve Location for FieldDoesNotExist exception?
- [ ] Add documentation for is_reverse_object (or remove completely?)
- [ ] Move related flags to ForeignKey, rather than ForeignObject
- [ ] Final flake8 check
- [ ] Final python -Wall runtests.py check

**Comment by freakboy3742** [comment]
@timgraham Regarding your point 3 about ForeignObject - GenericRelation does reverse some of the flags, and ForeignObject is undocumented - but it *is* tested, and if you take those flags out, some tests fail. So - my inclination is to leave the flags on ForeignObject as they are; yes, they're going to be altered by some subclasses, but they're reasonable defaults, and easy to change if the default value isn't right.

File: django/db/models/fields/related.py [hunk 3]
```diff
@@ -1273,26 +1281,46 @@ def __init__(self, field, to, related_name=None, limit_choices_to=None,
     # AttributeError
     @cached_property
     def model(self):
-        if not self.field.model:
-            raise AttributeError(
```
Comment (timgraham) [inline django/db/models/fields/related.py]: Just want to make sure this was a conscious deletion and not a bad merge conflict cause it was from the `RelatedObject` removal and not testing I think.

File: django/db/models/fields/related.py [hunk 4]
```diff
@@ -1318,7 +1346,7 @@ def get_db_prep_lookup(self, lookup_type, value, connection, prepared=False):
 
     def is_hidden(self):
         "Should the related object be hidden?"
-        return self.related_name and self.related_name[-1] == '+'
+        return bool(self.related_name and self.related_name[-1] == '+')
```
Comment (timgraham) [inline django/db/models/fields/related.py]: Just wondering why the `bool()` addition was needed here?

File: tests/invalid_models_tests/test_relative_fields.py [hunk 5]
```diff
@@ -47,6 +47,8 @@ def test_many_to_many_to_missing_model(self):
         class Model(models.Model):
             m2m = models.ManyToManyField("Rel2")
 
+        # Calling include_related=False because, as this is an invalid model
```
Comment (timgraham) [inline tests/invalid_models_tests/test_relative_fields.py]: Are these comments still relevant? I don't see anything with `include_related=False` here.

File: tests/lookup/tests.py [hunk 6]
```diff
@@ -472,7 +472,7 @@ def test_error_messages(self):
             self.fail('FieldError not raised')
         except FieldError as ex:
             self.assertEqual(str(ex), "Cannot resolve keyword 'pub_date_year' "
-                             "into field. Choices are: author, author_id, headline, "
```
Comment (timgraham) [inline tests/lookup/tests.py]: Not sure `author_id` should be removed from this list.

File: tests/m2m_and_m2o/tests.py [hunk 7]
```diff
@@ -5,6 +5,13 @@
 
 
 class RelatedObjectTests(TestCase):
+
+    def test_related_objects_have_name_attribute(self):
+        for Model in (Issue, User, UnicodeReferenceModel,):
```
Comment (timgraham) [inline tests/m2m_and_m2o/tests.py]: It would be better to structure this test to pull out the expected fields and assert each of them individually. That way if the test fails it'll be easier to debug, and also you ensure that some assertions are actually made (some changes to the model might make it so that to fields pass `is_reverse_object` and then the test does nothing).

**Comment by timgraham** [comment]
Changes, mostly cosmetic: http://dpaste.com/2WDN0F1
There is one TODO that needs to be addressed in the docs however.

Other notes:
* I didn't follow the discussion on the mailing list closely, but there are some places that now read `related.related_model` as opposed to `related.model`. Just wanted to double check that "the related model of a related object" as the code now indicates is more accurate description of what should be there.
* Should we name `model_meta/test.py` -> `tests.py` for consistency with the rest of the test packages?

Two things to add to the checklist:
* Investigate why the tracebacks on ref/models/meta.html render twice.
* squash commits

File: tests/lookup/tests.py (hunk 6 above)
Comment (collinanderson) [inline tests/lookup/tests.py]: I bet it's because author_id will now be a hidden field, so the question is: should hidden fields appear in this list?

File: tests/invalid_models_tests/test_relative_fields.py (hunk 5 above)
Comment (freakboy3742) [inline tests/invalid_models_tests/test_relative_fields.py]: Correct - this is a reference to a removed argument. Fixed.

**Comment by PirosB3** [comment]
buildbot, test this please.

File: django/db/models/fields/related.py (hunk 3 above)
Comment (PirosB3) [inline django/db/models/fields/related.py]: @timgraham yes this is intentional. This was needed more on FieldFlagsMixin.

File: django/db/models/fields/related.py (hunk 4 above)
Comment (PirosB3) [inline django/db/models/fields/related.py]: because is_hidden should always be a boolean flag (related_name can be None). Refactored it to look a bit nicer (removing the boolean cast). let me know if there are problems

**Comment by freakboy3742** [comment]
@timgraham Can you elaborate on how/where you're seeing duplicated tracebacks in ref/models/meta.html? I'm rendering the docs locally, and not seeing any duplicates.

I'm also not sure what you're referring to about the TODO in the docs - I can't find any todo markers, and there isn't any obvious dangling text to be addressed.

Other notes:
* I've renamed `model_meta/test.py` for consistency
* Closing the loop on the ForeignObject discussion - on IRC, @timgraham deferred judgement, so let's leave the flags on ForeignObject.
* The `related.model` vs `related_model` thing is a part of the refactor that came from removing `parent_model`. There will sometimes be multiple paths to get to a model (since both "ends" of a relation know the model and the related model); requests for related.related_model could probably be formulated in other ways, but it's just as valid. There are also some backwards incompatibilities around these changes, but then, there are already a bunch of those around the removal of `RelatedObject`. I'll do a final audit to make sure that `model` and `related_model` are all set correctly, but otherwise, I don't think there's anything needing attention.

**Comment by freakboy3742** [comment]
I've just audited the `related_model`/`model` code, and  - essentially `rel.related_model == rel.field.model`, `rel. model == rel.field.related_model`. This is the price we pay for being able to query `related_model` and `model` on any relation-like object. 

There is a small bug with reference resolution (string references to models aren't being resolved correctly), but the right models are being referenced. I'll clean up the string references (item added to TODO list), but I think we can close off this issue.

**Comment by freakboy3742** [comment]
@PirosB3 Did you ever open a ticket (or find an existing ticket) for this feature? I seem to recall there was one, but I can't find it in Trac.

**Comment by PirosB3** [comment]
buildbot, test this please.

**Comment by PirosB3** [comment]
@freakboy3742 is this a good ticket? https://code.djangoproject.com/ticket/12663

**Comment by PirosB3** [comment]
@freakboy3742 @carljm looking at removing local_* in base.py

**Comment by PirosB3** [comment]
ok @carljm @freakboy3742 unfortunately looks like we can't just replace local_* calls in models/base.py. The reason is because, as get_fields() call the relation tree, apps is obviously not ready at that point.
We can, however, replace local_* calls to _get_fields() (the internal API). Any thoughts?

**Comment by PirosB3** [comment]
RE removing is_reverse_object.
I have just being doing an evaluation on this, and the answer is: yes, we can remove is_reverse_object but only with one additional flag: this flag is called “auto_created”. “auto_created" basically means that the field is created dynamically and is not defined by the user. This flag is far more generic than "is_reverse_object" and I think it can benefit both externally and internally.

Update: "auto_created" already exists, we need to find another, better, name.

**Comment by PirosB3** [comment]
buildbot, test this please.

**Comment by PirosB3** [comment]
@freakboy3742 nice! so it looks like ```is_reverse_object``` has been removed from the code base! This has been possible by using the already existing ```auto_created``` flag.

File: django/core/serializers/python.py [hunk 8]
```diff
@@ -101,12 +101,13 @@ def Deserializer(object_list, **options):
         if 'pk' in d:
             data[Model._meta.pk.attname] = Model._meta.pk.to_python(d.get("pk", None))
         m2m_data = {}
-        model_fields = Model._meta.get_all_field_names()
+
+        field_names = set(f.name for f in Model._meta.get_fields())
```
Comment (collinanderson) [inline django/core/serializers/python.py]: minor thing, but you could use a set literal here like below

File: django/db/models/base.py [hunk 9]
```diff
@@ -1438,12 +1440,17 @@ def _check_unique_together(cls):
     def _check_local_fields(cls, fields, option):
         from django.db import models
 
+        # In order to avoid hitting the relation tree prematurely, we use our own
+        # fields_map instead of using get_field()
+        forward_fields_map = dict(
+            (field.name, field) for field in cls._meta._get_fields(reverse=False)
+        )
```
Comment (collinanderson) [inline django/db/models/base.py]: again, pretty minor, but this could be a dict literal

File: django/db/models/deletion.py [hunk 10]
```diff
@@ -51,6 +52,23 @@ def DO_NOTHING(collector, field, sub_objs, using):
...
+    candidate_models = candidate_models.union(opts.concrete_model._meta.proxied_children)
+    # For each model, get all candidate fields.
+    candidate_model_fields = chain.from_iterable(
+        opts.get_fields(include_hidden=True) for opts in candidate_models
+    )
+    # Keep related objects not coming from ManyToManyFields from all candidate
+    # model fields.
+    return (
+        f for f in candidate_model_fields
+        if f.auto_created and not f.concrete and not f.many_to_many
```
Comment (collinanderson) [inline django/db/models/deletion.py]: I could imagine non-auto_created reverse relations. Could we do something like: `if f.one_to_one or f.one_to_many and not f.concrete`?

File: django/db/models/fields/related.py [hunk 11]
```diff
@@ -1741,6 +1778,12 @@ def contribute_to_related_class(self, cls, related):
 
 
 class ForeignKey(ForeignObject):
+    # Field flags
+    one_to_many = True
```
Comment (collinanderson) [inline django/db/models/fields/related.py]: umm.... am I crazy? Aren't ForeignKeys many_to_one? and the reverse of a ForeignKey is one_to_many?

Right? if Book has ForeignKey(Author), then many Books have one Author and one Author has many Books?

File: django/db/models/fields/related.py [hunk 12]
```diff
@@ -2036,6 +2084,12 @@ def set_managed(field, model, cls):
 
 
 class ManyToManyField(RelatedField):
+    # Field flags
+    one_to_many = False
+    one_to_one = False
+    many_to_many = True
+    many_to_one = False
```
Comment (collinanderson) [inline django/db/models/fields/related.py]: can all of these be set to `False` on `RelatedField`?

File: django/contrib/admin/options.py [hunk 13]
```diff
@@ -473,9 +473,12 @@ def to_field_allowed(self, request, to_field):
             for inline in admin.inlines:
                 registered_models.add(inline.model)
 
-        for related_object in (opts.get_all_related_objects(include_hidden=True) +
-                               opts.get_all_related_many_to_many_objects()):
-            related_model = related_object.model
+        related_objects = (
+            f for f in opts.get_fields(include_hidden=True)
+            if (f.auto_created and not f.concrete)
```
Comment (collinanderson) [inline django/contrib/admin/options.py]: like above, would it work to do: `f.one_to_one and f.one_to_many and not f.concrete`?

File: django/contrib/admin/utils.py [hunk 14]
```diff
@@ -424,19 +435,21 @@ def reverse_field_path(model, path):
...
         # skip trailing data field if extant:
         if len(reversed_path) == len(pieces) - 1:  # final iteration
             try:
                 get_model_from_relation(field)
             except NotRelationField:
                 break
-        if direct:
+
+        # Field should point to another model
+        if field.has_relation and not (field.auto_created and not field.concrete):
```
Comment (collinanderson) [inline django/contrib/admin/utils.py]: again, it seems to me `auto_created` shouldn't matter in this case.

File: django/db/models/sql/query.py [hunk 15]
```diff
@@ -606,8 +613,11 @@ def deferred_to_data(self, target, callback):
                 if not is_reverse_o2o(source):
                     must_include[old_model].add(source)
                 add_to_dict(must_include, cur_model, opts.pk)
-            field, model, _, _ = opts.get_field_by_name(parts[-1])
-            if model is None:
+            field = opts.get_field(parts[-1])
+            is_reverse_object = field.auto_created and not field.concrete
+            model = field.related_model if is_reverse_object else field.model
```
Comment (collinanderson) [inline django/db/models/sql/query.py]: Again, it would be be great if we could handle non-auto-created `one_to_many` related fields. Would it work to do: `model = field.related_model or field.model` or the more explicit: `model = field.related_model if field.related_model else field.model`?

File: django/db/models/sql/query.py [hunk 16]
```diff
@@ -606,8 +613,11 @@ def deferred_to_data(self, target, callback):
                 if not is_reverse_o2o(source):
                     must_include[old_model].add(source)
                 add_to_dict(must_include, cur_model, opts.pk)
-            field, model, _, _ = opts.get_field_by_name(parts[-1])
-            if model is None:
+            field = opts.get_field(parts[-1])
+            is_reverse_object = field.auto_created and not field.concrete
```
Comment (collinanderson) [inline django/db/models/sql/query.py]: Again, it would be be great if we could handle non-auto-created `one_to_many` related fields. Not sure what this code does, but would it work to do:
`is_reverse_object = field.related_model and not field.concrete`?

File: django/db/models/fields/related.py (hunk 12 above)
Comment (freakboy3742) [inline django/db/models/fields/related.py]: They could be; however, IMHO there's also value in being explicit at the point where they're actually being used. I'm not too fussed either way, though.

File: django/db/models/fields/related.py (hunk 11 above)
Comment (freakboy3742) [inline django/db/models/fields/related.py]: I disagree - To my mind, ForeignKeys are One-to-many: they're "one" on the side that defines them, and "many" on the other side. This is backed up by the fact that the "other" side of a ForeignKey is a ManyToOneRel. 

It's worth noting that this same confusion exists in various [Entity-Relationship diagram notations](http://en.wikipedia.org/wiki/Entity–relationship_model#Cardinalities) - the side on which you put the visual "many" varies (c.f. Martin/Crow foot and IDEF1X notation). So, to some extent, this is a bikeshed; my preference would be to stick with the notation we've been using historically.

File: django/db/models/base.py (hunk 9 above)
Comment (freakboy3742) [inline django/db/models/base.py]: Fixed.

File: django/core/serializers/python.py (hunk 8 above)
Comment (freakboy3742) [inline django/core/serializers/python.py]: Fixed.

File: django/db/models/deletion.py (hunk 10 above)
Comment (freakboy3742) [inline django/db/models/deletion.py]: For me, something tied to `auto_created` makes more sense - what we're trying to capture here is whether the field is a canonical representation, or the "other" side that needs to exist for data access purposes, but doesn't require explicit action (because the action will happen on the canonical side). 

If we were to tie this to `one_to_one`/`one_to_many`, then we're assuming that a `one_to_one`/`one_to_many` field will always be the canonical representation requiring action. GenericRelations show how this isn't necessarily the case. 

A theoretical "OneToManyField" (i.e., a "reverse" definition of a ForeignKey) would be another interesting case - that's a field where the "concrete" column is on the reverse, auto-created object.

Comment (collinanderson) [inline django/db/models/deletion.py]: In this theoretical case, you would want the related objects to recursively delete, right? Shouldn't it be included in this list?

Also, wouldn't you call the "reverse" of a ForeignKey a ManyToOneField? (if a ForeignKey a "one to many")

File: django/contrib/admin/options.py (hunk 13 above)
Comment (PirosB3) [inline django/contrib/admin/options.py]: what about ```GenericForeignKey``` then? ;)
one-to-many CHECK!
not f.concrete CHECK!

In fact, there is very little difference from a ```ManyToOneRel``` reverse relation and a ```GenericReltion```!

- non concrete
- both have a reverse model
- same cardinality (m2o)
- by default, both non hidden,

The only thing that *really* differentiates the two, is that one is created by default (not user-defined) and the other one is. For this reason the flag ```auto_created``` is a key distinguisher.

**Comment by PirosB3** [comment]
buildbot, test this please.

**Comment by freakboy3742** [comment]
buildbot, test this please.

File: django/db/models/deletion.py (hunk 10 above)
Comment (freakboy3742) [inline django/db/models/deletion.py]: Firstly - my apologies, I meant ManyToOneField, not OneToManyField as my candidate example.

As for the proposed change: I'm of two minds. Any discussion around this topic involves a certain amount of architecture astronauting, because we're discussing the needs of a completely theoretical class of fields that doesn't exist in the Real World (AFAICT). 

@collinanderson's approach is explicit on the fields that we currently have, but will obviously be weaker *if* other field types are defined. @PirosB3's approach is stronger from a theoretical perspective, because it tries to capture the concept, rather than the specific, but we don't have anything to verify whether the concept is actually correct in practice, other than a ManyToOneField that doesn't exist. As @collinanderson points out, the ManyToOneField demonstrates that the currently logic isn't correct as written; I'm sure there's a more correct way of expressing the abstract idea, but it isn't immediately obvious to me.

So - we have three options AFAICT:
 * Keep it as is, because It Works(tm), and *might* be right for a certain class of extension fields.
 * Wring our hands trying to fix the logic for an theoretical case that doesn't exist.
 * Change to @collinanderson's suggestion, on the basis that it's more explicit, and treat any problems identified with extension fields as bugs to be addressed over time.

Thoughts?

Comment (loic) [inline django/db/models/deletion.py]: Doesn't `GenericRelation` qualify as an existing `ManyToOneField`? Also I recently discussed with @akaariai the possibility of making every reverse relation actual `Field` subclasses (see https://github.com/django/django/pull/2411#issuecomment-68508995 for some context), so that would imply creating a `ManyToOneField` field. Also a third party field could easily implement a `ManyToOneField` backed by a M2M with a unique constraint.

File: django/contrib/admin/options.py (hunk 13 above)
Comment (freakboy3742) [inline django/contrib/admin/options.py]: @PirosB3 There's one significant difference - a `GenericRelation` is backed by a `GenericForeignKey` definition on the other model; a theoretical `ManyToOneField` doesn't have anything concrete on the other end to provide the "actual" definition. `GenericRelation` objects can be ignored for anything other than data lookup purposes - anything to do with deletion etc can be ignored. `ManyToOneField` would need to invoke some functionality.

So - same argument as above - do we use what is there right now; try to fix it for an example that doesn't exist, or use a more explicit syntax that will be known to be incorrect in a theoretical case, but not for any actually extant case?

File: django/db/models/deletion.py (hunk 10 above)
Comment (freakboy3742) [inline django/db/models/deletion.py]: @loic - `GenericRelation` it's certainly a many-to-one relation, but it differs from a "native" ManyToOneField on the basis that `GenericRelation` is a lookup aid for `GenericForeignKey`, whereas a "native" ManyToOneField would be a field that actually adds a column to the related model. 

I'm not saying that a common base class can't/shouldn't be used - but there would need to be some way to tell the two cases apart. 

That said, It's difficult to account for cases where there isn't any actual code, which leads me to suspect the "explicit, but probably wrong for cases we haven't seen yet" (i.e., @collinanderson's suggestion) is probably the best approach.

Comment (loic) [inline django/db/models/deletion.py]: Thinking about it a bit more, I wouldn't rely on `concrete` either. Once we add composite FK support, `ForeignKey.concrete` will be `False`. Can't we just check for `if field.many_to_one or field.one_to_one`? Then further down the road we check if the relation is nullable (for which we'll probably need a higher level concept).

Comment (loic) [inline django/db/models/deletion.py]: @freakboy3742 I agree that "for now" `GenericRelation` is just a lookup aid, but ideally it shouldn't be just that. Eventually we want all reverse relation to be actual fields (note that in case of `GenericRelation` we may need to introduce a new field to preserve backwards compatibility for people who really just wanted "lookup aid"). If I made a M2M backed M2O to avoid tight coupling 2 apps (i.e. you can't control a third party app model), I'd like it to be included in deletion. It would also make sense in the context of a "reverse O2O" such as https://github.com/akaariai/django-reverse-unique/tree/master/reverse_unique.

Comment (freakboy3742) [inline django/db/models/deletion.py]: Discussion on IRC (with @freakboy3742, @loic and @collinanderson) came to he conclusion that option (3) - being explicit about current capabilities - was the best option. In future, we may need to add more flags, and those flags will inevitably alter the usage of filters in practice, but coding now in presumption of what might land will just end up with code that is harder to read right now, and almost certainly incorrect in future. 

@PirosB3 - do you agree?

File: django/apps/registry.py [hunk 17]
```diff
@@ -331,13 +331,19 @@ def unset_installed_apps(self):
         self.apps_ready = self.models_ready = self.ready = True
         self.clear_cache()
 
-    def clear_cache(self):
+    def clear_cache(self, expire_models_cache=False):
```
Comment (aaugustin) [inline django/apps/registry.py]: During the app-loading refactor I've tried to eliminate as much as possible methods that did something completely different depending on keyword arguments. Here you're only adding some new logic to clear_cache() hidden behind an `if` clause.

I think model caches should be expired in all cases. For instance, if I remove an app from installed apps, reverse relations to models defined in that app should be cleared from the caches of models defined in other apps. That's why I think you should simply drop the `expire_models_cache` argument and the `if` clause.

If you need to keep them for a reason I missed, can you make a separate method:

    def clear_models_cache(self):
        self.clear_cache()
        self.check_apps_ready()
        for model in self.get_models(include_auto_created=True):
            model._meta._expire_cache()

and replace `apps.clear_cache(expire_models_cache=True)` by `apps.clear_models_cache()`?

File: django/db/migrations/state.py [hunk 18]
```diff
@@ -201,7 +201,7 @@ def register_model(self, app_label, model):
             self.app_configs[app_label] = AppConfigStub(app_label)
             self.app_configs[app_label].models = OrderedDict()
         self.app_configs[app_label].models[model._meta.model_name] = model
-        self.clear_cache()
+        self.clear_cache(True)
```
Comment (aaugustin) [inline django/db/migrations/state.py]: In addition to my previous comment above, passing boolean flags as positional arguments should be avoided because the semantics is impossible to understand without checking the definition of the target function.

File: docs/internals/deprecation.txt [hunk 19]
```diff
@@ -56,6 +56,19 @@ details on these changes.
 
 * ``django.template.resolve_variable`` will be removed.
 
+* The following methods will be removed from
```
Comment (freakboy3742) [inline docs/internals/deprecation.txt]: Done.

File: docs/releases/1.8.txt [hunk 20]
```diff
@@ -29,6 +29,32 @@ Like Django 1.7, Django 1.8 requires Python 2.7 or above, though we
 What's new in Django 1.8
 ========================
 
+``Model._meta`` API
+~~~~~~~~~~~~~~~~~~~
+
+Django now has an official API for ``Model._meta``, providing an officially
```
Comment (freakboy3742) [inline docs/releases/1.8.txt]: Done.

File: docs/releases/1.8.txt [hunk 21]
```diff
@@ -990,6 +1016,126 @@ Miscellaneous
 Features deprecated in 1.8
 ==========================
 
+.. _model_meta_upgrade:
```
Comment (freakboy3742) [inline docs/releases/1.8.txt]: I've moved this to a section at the end of the Meta docs, because it didn't seem substantial enough to warrant an entirely new page, and that page already contains references to the fact that there was an "old" API.

File: docs/releases/1.8.txt [hunk 22]
```diff
@@ -29,6 +29,32 @@ Like Django 1.7, Django 1.8 requires Python 2.7 or above, though we
...
+supported way to :ref:`retrieve fields <model-meta-field-api>` and filter
+fields based on their :ref:`attributes <model-field-attributes>`.
+
+The ``Model._meta`` object has been part of Django since the days of pre-0.96
+"Magic Removal" -- it just wasn't an official, stable API. In recognition of
+this, we've endeavored to maintain backwards-compatibility with the old
+API endpoint where possible. However, API endpoints that aren't part of the
+new official API have been deprecated and will eventually be removed.
+
+The new API is composed of two methods:
```
Comment (freakboy3742) [inline docs/releases/1.8.txt]: Done.

File: django/apps/registry.py (hunk 17 above)
Comment (freakboy3742) [inline django/apps/registry.py]: The boolean was introduced because you can't expire the model cache in *all* cases - the apps registry needs to be ready before you can flush the model cache, because you have to interrogate the list of models to build the model cache. The apps registry does aggressive self-flushing every time a model is loaded, so this causes a problem.

However, if you just don't flush the model cache util "apps.ready", you can avoid this problem. This stands to reason AFAICT; you won't be doing any model operations until the app cache is populated, and the model cache is populated on first use, so there can't be a model cache until the apps cache is ready.

File: django/db/migrations/state.py (hunk 18 above)
Comment (freakboy3742) [inline django/db/migrations/state.py]: Agreed - all other uses have explicitly called `expire_models_cache=True`. However, this is now a moot point since we can remove `expire_models_cache`.

File: django/contrib/admin/options.py (hunk 13 above)
Comment (freakboy3742) [inline django/contrib/admin/options.py]: On closer inspection - I think @PirosB3's version actually *might* be the simplest correct version here. The original code is looking for all related_objects and related_m2m_objects -- that's effectively what `auto_created` now covers - `auto_created` is being used in a lot of places `is_related_field` was being used previously. You can't *just* use `one_to_one` and `many_to_one`, because that misses `many_to_many`; you can't just use `not concrete`, because of `GenericRelation`. You have to use both `auto_created and not concrete` because of `AutoField`

We *could* put in a `many_to_many or many_to_one or one_to_one` clause for the sake of clarity, but that isn't strictly necessary based on the fields we have right now.

File: django/db/models/deletion.py (hunk 10 above)
Comment (freakboy3742) [inline django/db/models/deletion.py]: On closer inspection - I think @PirosB3's version is pretty close to the simplest correct version here. We're looking for all related_objects that haven't come from a m2m. We can't *just* use `many_to_one and one_to_one` - we need auto_created to make sure we exclude GenericRelation; we also need `not concrete` to exclude proxies.

The only alternate phrasing of this clause that I can see working is:
```
if f.auto_created and not f.concrete and (f.one_to_one or f.many_to_one)
```
which is slightly more explicit about the fields it wants than `not many_to_many`, so I've made that change.

File: django/contrib/contenttypes/fields.py [hunk 23]
```diff
@@ -21,19 +21,33 @@ class GenericForeignKey(object):
...
+    many_to_many = False
+    many_to_one = False
+    one_to_one = False
+    related_model = None
 
     def __init__(self, ct_field="content_type", fk_field="object_id", for_concrete_model=True):
         self.ct_field = ct_field
         self.fk_field = fk_field
         self.for_concrete_model = for_concrete_model
+        self.generate_reverse_relation = False
```
Comment (timgraham) [inline django/contrib/contenttypes/fields.py]: `generate_reverse_relation` seems unused now? (also defined in `db/models/fields/related.py`)

File: django/contrib/contenttypes/fields.py [hunk 24]
```diff
@@ -21,19 +21,33 @@ class GenericForeignKey(object):
...
         self.editable = False
-        self.rel = None
+        self.rel = True
```
Comment (timgraham) [inline django/contrib/contenttypes/fields.py]: On normal fields, `self.rel` is either `None` or something like `ManyToOneRel` -- is `True` really also an acceptable value? This seems slightly confusing (especially when it was `None` before). It seems like a boolean would be more for something like `is_relation`, but not sure on this since `rel` isn't a public attribute. Maybe a clarifying comment is at least in order.

File: django/db/migrations/autodetector.py [hunk 25]
```diff
@@ -685,26 +685,15 @@ def generate_deleted_models(self):
...
-                    related_object.field.name,
-                    False,
-                ))
+            for related_object in model._meta.related_objects:
+                related_object_app_label = related_object.related_model._meta.app_label
+                object_name = related_object.related_model._meta.object_name
+                field_name = related_object.field.name
+
+                dependencies.append((related_object_app_label, object_name, field_name, False))
+                if not isinstance(related_object, models.ManyToManyRel):
```
Comment (timgraham) [inline django/db/migrations/autodetector.py]: is there a reason we aren't using field flags here?

File: django/db/models/fields/related.py (hunk 3 above)
Comment (timgraham) [inline django/db/models/fields/related.py]: Not sure I understand that reply since `FieldFlagsMixin` no longer exists.

File: django/db/models/fields/related.py (hunk 4 above)
Comment (timgraham) [inline django/db/models/fields/related.py]: Seems like you only removed the cast in the `is_hidden()` method below in this file.

File: django/db/models/fields/related.py [hunk 26]
```diff
@@ -1954,6 +1997,11 @@ class OneToOneField(ForeignKey):
     related_accessor_class = SingleRelatedObjectDescriptor
     description = _("One-to-one relationship")
 
+    many_to_many = False
```
Comment (timgraham) [inline django/db/models/fields/related.py]: should these be listed above the other attributes of the class for consistency with the other fields? Also missing `# Field flags` comment like the others.

File: django/db/models/manager.py [hunk 27]
```diff
@@ -23,15 +22,12 @@ def ensure_default_manager(cls):
         setattr(cls, 'objects', SwappedManagerDescriptor(cls))
         return
     if not getattr(cls, '_default_manager', None):
-        # Create the default manager, if needed.
-        try:
-            cls._meta.get_field('objects')
+        if any(f.name == 'objects' for f in cls._meta.fields):
```
Comment (timgraham) [inline django/db/models/manager.py]: Is this change needed? I'm wondering why iterating over `fields` is preferred to `get_field()` as it seems like the new way would be slower and is a bit less explicit IMO.

Comment (PirosB3) [inline django/db/models/manager.py]: Good question: get_fields() necessarily warms the relation tree cache for everyone, we don’t want this to happen before the apps registry is ready.

**Comment by timgraham** [comment]
Should Field.auto_created should be documented in "Attributes for fields"?

**Comment by PirosB3** [comment]
@timgraham RE  Field.auto_created:
There already exists a documentation for auto_created, I think it would be beneficial to move it, or at least to link it there. What do you think is the most appropriate solution?

File: django/db/models/manager.py (hunk 27 above)
Comment (PirosB3) [inline django/db/models/manager.py]: added a comment to explain

File: django/contrib/contenttypes/fields.py (hunk 23 above)
Comment (PirosB3) [inline django/contrib/contenttypes/fields.py]: removed

File: django/db/models/fields/__init__.py [hunk 28]
```diff
@@ -59,9 +59,13 @@ class NOT_PROVIDED:
 # of most "choices" lists.
 BLANK_CHOICE_DASH = [("", "---------")]
 
+# When the _meta object was formalized, this exception was moved to
+# django.core.exceptions. It is retained here for backwards compatibility
+# purposes.
+
```
Comment (PirosB3) [inline django/db/models/fields/__init__.py]: Thanks @pipermerriam I think this was my fault!

File: docs/ref/models/meta.txt [hunk 29]
```diff
@@ -0,0 +1,277 @@
...
+
+  - ``f.auto_created == False``, because the new ``get_field()``
+    API will find "reverse" relations), and:
+
+  - ``f.is_relation and f.related_model is None``, because the new
+    ``get_field()`` API will find
+    :class:``~django.contrib.contenttypes.fields.GenericForeignKey``
+    relations;
+
+* ``MyModel._meta.get_field_by_name(name)[0]``::
```
Comment (timgraham) [inline docs/ref/models/meta.txt]: It may be useful to address the case where someone doesn't just index the first item of the tuple.

**Comment by timgraham** [comment]
merged in fb48eb05816b1ac87d58696cdfe48be18c901f16
